### ISSUE TITLE
Speech2text demo results way off

### ISSUE THREAD

**Issue by kaihuchen** [OpenNMT/OpenNMT-py#808]
I am trying to duplicate the speech2text demo as described here: http://opennmt.net/OpenNMT-py/speech2text.html

I am running the demo using the repository code cloned just today, the demo runs well without problem (except that I had to hack a little into onmt/utils/optimizers.py around line #215 in order to bypass an unrelated error about self.start_decay_steps).  The training phase converges to 100% accuracy and 1.00 perplexity fairly quickly:

    ...Step 34950, 53000; acc: 100.00; ppl:   1.00; xent:   0.00; lr: 0.00111;   0 / 3023 tok/s;    406 sec

Testing the result with translate.py got me this result:

    PRED AVG SCORE: -0.0438, PRED PPL: 1.0448 

Comparing the translation with the expected text (i.e., data\speech\an4_dataset\test\an4\txt\*.txt) shows that (ignoring the extra spaces in the translation) the result is not good. 

I understand that this speech2text feature is relatively new so there could be some kinks. I just want to know whether I am seeing the expected result at this stage, or whether I made mistakes somewhere. Thanks!

**Comment by da03** [comment]
Thanks for reporting the poor generalization ability! What's the validation performance? Actually this An4 dataset is just a demo dataset and its size does not support training a pure seq2seq-attn model. (For baidu deepspeech2 they used CTC loss which can train with much less data though cannot learn a language model). I'd suggest looking at this PR `https://github.com/OpenNMT/OpenNMT-py/pull/759` if you are interested in training on a real dataset since 1) it supports sharding which is necessary for preprocessing large datasets, and 2) it's been tested on a much larger Libri-speech dataset.

**Comment by kaihuchen** [comment]
@da03  Thanks for sharing info about the sharding feature. This should come really handy when we are ready to go full speed with training a large dataset.

Regarding the validation error, I restarted the training process with valid_steps set to 100. After 100000 steps of training I got the validation accuracy at around 81%, as follows:

Step 100000, 100000; acc: 100.00; ppl:   1.00; xent:   0.00; lr: 0.10000;   0 / 2385 tok/s;   7103 sec
Loading valid dataset from demo.valid.pt, number of examples: 130
Validation perplexity: 7.22387
Validation accuracy: 81.41

The validation accuracy changes very little after 25000 steps.

I understand that with only 943 training samples, this demo is not meant for learning a full language model. What puzzles me is that with training accuracy at 100% and validation accuracy at ~81% (which suggests perhaps some degree of over-fitting, but overall not too bad), I am expecting the test result to be much better than what I have observed.

**Comment by da03** [comment]
Hmm validation ppl of 7.22 and accuracy of 81% is actually bad (vocab size is only 37 or so, i suspect even a pure language model would get a decent performance on this task), since the evaluation is by feeding the gold sequence prefix and calculate if the current predicted probabilities get a maximum at the correct token, which is not what we do at test time (we are feeding model's prediction rather than ground truth prefix).

**Comment by kaihuchen** [comment]
I was still unable to get good result out of the speech demo, so I try it again with my own dataset to test it. What I have is a small dataset of 1319 samples splitted 7-1-2 for training/validation/testing. The source samples are synthesized speech of Chinese numbers with only 13 tokens, the target samples are phonetic representation (called pinyin) of the source speech, where there is a simple one-to-one perfect source-target mapping. And the training result was still pretty bad even though training accuracy/perplexity is reported as 100/1.0, and validation's is 99.9/1.01 when I cut off the training.  

As a test, I applied the trained model on the training dataset expecting to see perfect results, but it was far from it. Then I noticed that there is actually a perfect PARTIAL MATCH, in the sense that the tokens in the predicted target is always a PROPER SUBSET of the tokens in the true target. That leads me to wonder if we need to adjust the loss function in some way.

**Comment by da03** [comment]
Hmm that's weird. Any chance you can share me the model and a few training examples (such as by google drive or dropbox)? Which version of code are you using?

**Comment by kaihuchen** [comment]
Here they are: https://drive.google.com/drive/folders/1gqAunarvrxQT0uYTJxRL6Kw7yMG1_yFZ

In there you will find the following:
1. Five wave files that I randomly selected from the training sources.
2. Three data pt files:  dataset.train.pt, dataset.valid.pt, dataset.vocab.pt
3. The trained model: model_latest.pt
4. The source data files for training, which I eventually used for testing (as described above): train_src-full.txt, train_tgt-full.txt
5. train-inferred.txt, which contains the predictions made with model_latest.pt based on train_src-full.txt as the input. If you diff the two files train-inferred.txt and train_tgt-full.txt (the ground truth prediction), then you will see the 'proper subset' phenomenon that I mentioned above.

I am using the code that I cloned from github on July 8th.

Hope this helps!

**Comment by kaihuchen** [comment]
@da03  Regarding the sharding feature you mentioned above, I see that that pull issue #759 seems still  ongoing. Do you know whether it is going to be available soon? I have a fairly large (>5GB) audio training dataset that does not run at all (the training process invariable die by itself), and I figure that this sharding feature should help greatly.

**Comment by da03** [comment]
Hi @kaihuchen it's actually ready for use now (I have trained a model on librispeech and am pretty sure it's working), although in order to merge that to latest OpenNMT more work needs to be done. I'll merge this pr asap, but I think you can try that pull issue for now.

**Comment by kaihuchen** [comment]
@da03   FYI, I managed to get near perfect result with my speech-to-pinyin example using your new code. For some unknown reasons I have not been able to get the same level of test result with the master branch of OpenNMT-py, even though it is reporting that the training has reached 100% accuracy and 1.0 perplexity. 

Thanks for the excellent job with making the sharding feature and the Baidu engine available. I will proceed to try it on more complex tasks now.

**Comment by da03** [comment]
Great to know that!
